File: _dsv/uniq.py
import argparse
from ._column_slicer import _ColumnSlicer
from . import _utils

class uniq(_ColumnSlicer):
    ''' omit repeated lines '''
    parser = argparse.ArgumentParser()
    parser.add_argument('fields', nargs='*', help='check these only fields for uniqueness')
    parser.add_argument('-x', '--complement', action='store_true', help='exclude, rather than include, field names')
    parser.add_argument('-r', '--regex', action='store_true', help='treat fields as regexes')
    parser.add_argument('-c', '--count', action='store_true', help='prefix lines by the number of occurrences')
    parser.add_argument('-C', '--count-column', type=_utils.utf8_type, help='name of column to put the count in')
    group = parser.add_mutually_exclusive_group()
    group.add_argument('--group', action='store_true', help='show all items, separating groups with an empty line')
    group.add_argument('--repeated', action='store_true', help='only print duplicate lines, one for each group')
    group.add_argument('--repeated-all', action='store_true', help='print all duplicate lines')

    def __init__(self, opts):
        super().__init__(opts)
        opts.count_column = opts.count_column or (opts.count and b'count') or None
        self.uniq = {}
        self.counts = {}
        self.gather = opts.group or opts.repeated_all
        self.repeated = opts.repeated or opts.repeated_all
        self.print_early = not self.gather and opts.count_column is None
        self.sep = self.get_separator()

    def on_header(self, header):
        if self.opts.count_column is not None:
            header = [self.opts.count_column] + header
        return super().on_header(header)

    def on_row(self, row):
        key = tuple(self.slice(row, self.opts.complement))

        count = self.counts[key] = self.counts.get(key, 0) + 1
        rows = self.uniq.setdefault(key, [])
        if self.print_early:
            if count == (2 if self.repeated else 1):
                return super().on_row(row)
        elif self.gather or not rows:
            rows.append(row)

    def print_groups(self):
        for i, (k, rows) in enumerate(self.uniq.items()):
            count = self.counts[k]
            if self.repeated and count < 2:
                continue
            if self.opts.group and i > 0 and super().on_row(self.sep):
                return
            for row in rows:
                if self.opts.count_column:
                    row.insert(0, b'%i' % count)
                if super().on_row(row):
                    return

    def on_eof(self):
        if not self.print_early:
            self.print_groups()
        super().on_eof()

File: src/subcommands/uniq.rs
use anyhow::Result;
use crate::base;
use bstr::{BString};
use std::collections::{HashMap};
use crate::column_slicer::ColumnSlicer;
use clap::{Parser, ArgAction};

#[derive(Parser)]
#[command(about = "omit repeated lines")]
pub struct Opts {
    #[arg(help = "sort based only on these fields")]
    fields: Vec<String>,
    #[arg(short = 'x', long, action = ArgAction::SetTrue, help = "exclude, rather than include, field names")]
    complement: bool,
    #[arg(long, action = ArgAction::SetTrue, help = "treat fields as regexes")]
    regex: bool,
    #[arg(short = 'c', long, action = ArgAction::SetTrue, help = "prefix lines by the number of occurrences")]
    count: bool,
    #[arg(short = 'C', long, help = "name of column to put the count in")]
    count_column: Option<String>,
    #[arg(long, action = ArgAction::SetTrue, help = "show all items, separating groups with an empty line")]
    group: bool,
    #[arg(long, conflicts_with_all = ["group"], help = "only print duplicate lines, one for each group")]
    repeated: bool,
    #[arg(long, conflicts_with_all = ["group"], help = "print all duplicate lines")]
    repeated_all: bool,
}

type Rope = Vec<BString>;
pub struct Handler {
    map: HashMap<Rope, (usize, Vec<Rope>)>,
    gather: bool,
    repeated: bool,
    print_early: bool,
    column_slicer: ColumnSlicer,
    opts: Opts,
}

impl Handler {
    pub fn new(mut opts: Opts, _base: &mut base::Base) -> Result<Self> {
        if opts.count {
            opts.count_column.get_or_insert_with(|| "count".into());
        }

        let gather = opts.group || opts.repeated_all;
        Ok(Self {
            map: HashMap::new(),
            gather,
            repeated: opts.repeated || opts.repeated_all,
            print_early: !gather && opts.count_column.is_none(),
            column_slicer: ColumnSlicer::new(&opts.fields, opts.regex),
            opts,
        })
    }
}

impl base::Processor for Handler {
    fn on_header(&mut self, base: &mut base::Base, mut header: Vec<BString>) -> Result<bool> {
        self.column_slicer.make_header_map(&header);
        if let Some(count_column) = &self.opts.count_column {
            header.insert(0, count_column.as_bytes().into());
        }
        base.on_header(header)
    }

    fn on_row(&mut self, base: &mut base::Base, row: Vec<BString>) -> Result<bool> {
        let key = self.column_slicer.slice(&row, self.opts.complement, true);

        let entry = self.map.entry(key).or_insert((0, vec![]));
        entry.0 += 1;

        if self.print_early {
            if entry.0 == (if self.repeated { 2 } else { 1 }) {
                return base.on_row(row)
            }
        } else if self.gather || entry.1.is_empty() {
            entry.1.push(row);
        }

        Ok(false)
    }

    fn on_eof(self, base: &mut base::Base) -> Result<bool> {
        if !self.print_early {
            let mut first = true;
            'outer: for (_, (count, rows)) in self.map {
                if self.repeated && count < 2 {
                    continue
                }
                if self.opts.group && !first && base.on_separator() {
                    break
                }
                first = false;
                for mut row in rows {
                    if self.opts.count_column.is_some() {
                        row.insert(0, format!("{count}").into());
                    }
                    if base.on_row(row)? {
                        break 'outer
                    }
                }
            }
        }

        base.on_eof()
    }
}
